### PR TITLE
add the keyword recognizer

### PR DESCRIPTION
**Pull request by brentspell**
Most of this code is cloned from the wakeword trigger, but there probably isn't enough in common there to warrant trying to reuse any of it.

### CODE REVIEW

File: src/main/java/io/spokestack/spokestack/asr/KeywordRecognizer.java [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+ * using the VAD/timeout (or other activator) to manage activations.
+ * Alternatively, the recognizer can be used along with a wakeword detector
+ * to manage activations, in a two-stage wakeword/recognizer pattern.
+ * </p>
+ *
+ * <p>
+ * This pipeline component supports the following configuration properties:
+ * </p>
+ * <ul>
+ *   <li>
```
Comment (space-pope) [inline src/main/java/io/spokestack/spokestack/asr/KeywordRecognizer.java:66]: missing `keyword-classes`

File: src/main/java/io/spokestack/spokestack/asr/KeywordRecognizer.java [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+ *      "detect" Tensorflow-Lite model; its inputs shoudld be shaped
+ *      [encode-length, encode-width], and its outputs [len(classes)]
+ *   </li>
+ *   <li>
+ *      <b>keywod-pre-emphasis</b> (double): the pre-emphasis filter weight
+ *      to apply to the audio signal (0 for no pre-emphasis)
+ *   </li>
+ *   <li>
+ *      <b>keyword-fft-window-size</b> (integer): the size of the signal
+ *      window used to calculate the STFT, in number of samples - should be a
```
Comment (space-pope) [inline src/main/java/io/spokestack/spokestack/asr/KeywordRecognizer.java:90]: actually _has to be_ a power of 2, or an error is thrown

File: src/main/java/io/spokestack/spokestack/asr/KeywordRecognizer.java [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+
+    private void filter(SpeechContext context) {
+        // decode the FFT outputs into the filter model's inputs
+        // . compute the magnitude (abs) of each complex stft component
+        // . the first and last stft components contain only real parts
+        //   and are stored in the first two positions of the stft output
+        // . the remaining components contain real/imaginary parts
+        this.filterModel.inputs(0).rewind();
+        this.filterModel.inputs(0).putFloat(this.fftFrame[0]);
+        for (int i = 1; i < this.fftFrame.length / 2; i++) {
```
Comment (space-pope) [inline src/main/java/io/spokestack/spokestack/asr/KeywordRecognizer.java:366]: I know this loop is identical in the wakeword code, but isn't it more conventional to loop until `this.fftFrame.length` and increment by 2 on each iteration than to do all the inline multiplication?

File: src/main/java/io/spokestack/spokestack/asr/KeywordRecognizer.java (hunk 1 above)
Comment (brentspell) [reply src/main/java/io/spokestack/spokestack/asr/KeywordRecognizer.java:66]: nice, good catch

File: src/main/java/io/spokestack/spokestack/asr/KeywordRecognizer.java [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+ *      in vector units
+ *   </li>
+ *   <li>
+ *      <b>keyword-state-width</b> (integer): the size of the encoder state,
+ *      in vector units (defaults to keyword-encode-width)
+ *   </li>
+ *   <li>
+ *      <b>keyword-threshold</b> (double): the threshold of the classifier's
+ *      posterior output, above which the recognizer raises a recognition
+ *      event for the most likely kewyord class, in the range [0, 1]
```
[the remark this answers is not in the document]
Comment (brentspell) [reply src/main/java/io/spokestack/spokestack/asr/KeywordRecognizer.java:126]: yeah, I hate to write documentation for these when the defaults are constants in the module that you can see in the javadoc, too easy to get out of sync

File: src/main/java/io/spokestack/spokestack/asr/KeywordRecognizer.java (hunk 3 above)
Comment (brentspell) [reply src/main/java/io/spokestack/spokestack/asr/KeywordRecognizer.java:366]: I don't think there is a java convention here, we'll leave it consistent with the wakeword detector for now

**Comment by brentspell** [comment]
please squash for the typos/docstring fixes

**Pull request merged**